### PR TITLE
Release 4.5

### LINKED ISSUE #3: Move from `dep` to Go modules
https://github.com/jim-minter/rp/issues/4

### PR DESCRIPTION
**Pull request by mjudeikis**
Release 4.5 for ARO
* move to `go mod`
* Architecture differences 
    Single NSG for all components.
    Single Public loadbalancer for all components
* cgmanifest refactored to use gomod

TODO:
- [x] refactor installer/release-4.5-azure to sanitise commits [MJ] 
- [x] review usage of context.TODO(), context.Background() [MJ]
- [x] verify DNS migration code in installer (no-op for ARO) [MJ]
- [x] review go.mod settings for sanity - validate we're bringing in the right releases of things [JM]
- [x] build a public and a private cluster and check the LB configs with a toothcomb
- [ ] another careful code review

Installer updated fork: https://github.com/jim-minter/installer/commits/release-4.5-azure 

fixes #3 !!!

### CODE REVIEW

**Comment by mjudeikis** [comment]
@jim-minter  this is still needs polishing, but it is getting closer. I think first round of reviews can be done. And branch works to create new clusters too.

File: hack/vendor/vendor.go [hunk 1]
```diff
@@ -0,0 +1,250 @@
+package main
```
Comment (jim-minter) [inline hack/vendor/vendor.go:1]: I don't understand this file, I need it to be explained to me :-/
Also note: https://pkg.go.dev/golang.org/x/mod/modfile

File: Makefile [hunk 2]
```diff
@@ -111,4 +111,9 @@ test-python: generate pyenv${PYTHON_VERSION}
 admin.kubeconfig:
 	hack/get-admin-kubeconfig.sh /subscriptions/${AZURE_SUBSCRIPTION_ID}/resourceGroups/${RESOURCEGROUP}/providers/Microsoft.RedHatOpenShift/openShiftClusters/${CLUSTER} >admin.kubeconfig
 
-.PHONY: admin.kubeconfig aro az clean client generate image-aro image-fluentbit image-proxy image-routefix proxy publish-image-aro publish-image-fluentbit publish-image-proxy publish-image-routefix secrets secrets-update e2e.test test-e2e test-go test-python
+vendor:
```
Comment (jim-minter) [inline Makefile:114]: I wonder if we want to use `go build -mod=vendor` in this file to enforce that deps that are not checked in to vendor are caught?
General discussion needed around if/how we want to continue to use vendor/
- Protection against go module serving problems? cf https://github.com/gomodules/jsonpatch/issues/28
- This PR seems to mess up my "jump to definition" tooling in vs code, not sure why

File: docs/publish-rhcos-image.md [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+Each release we need to re-publish RHCOS image into azure cloud partners portal.
+
+1. Once new installer is vendored run `make vendor` to update local vendor directory.
+
+1. Run `make generate` to update generated content
+
+1. Run `go run ./hack/rhcos-sas/rhcos-sas.go` to copy RHCOS image. This might take a while.
+
+1. Command above will output SAS URL. Publish it via (partner)[https://partner.microsoft.com/] portal.
+If you need images/icons for the offer, you can find them in `docs/img`.
```
Comment (jim-minter) [inline docs/publish-rhcos-image.md:12]: I don't love this location but I don't have an obvious better suggestion

File: docs/publish-rhcos-image.md [hunk 4]
```diff
@@ -0,0 +1,12 @@
+# Publish RHCOS image
+
+Each release we need to re-publish RHCOS image into azure cloud partners portal.
+
+1. Once new installer is vendored run `make vendor` to update local vendor directory.
+
+1. Run `make generate` to update generated content
```
Comment (jim-minter) [inline docs/publish-rhcos-image.md:7]: `make vendor` does this for you

File: docs/publish-rhcos-image.md [hunk 5]
```diff
@@ -0,0 +1,12 @@
+# Publish RHCOS image
+
+Each release we need to re-publish RHCOS image into azure cloud partners portal.
```
Comment (jim-minter) [inline docs/publish-rhcos-image.md:3]:
```suggestion
Each release we need to re-publish the RHCOS image into the Azure cloud partner portal.
```

File: pkg/cluster/deployresources_resources.go [hunk 6]
```diff
@@ -0,0 +1,553 @@
+package cluster
```
Comment (jim-minter) [inline pkg/cluster/deployresources_resources.go:1]: I'm OK with adding this file, but it obfuscates any changes made as part of 4.4->4.5 so review is hard.  Please can you rebase and split the relevant commit into two: one which makes a no-op change, the other which has any changes?

File: pkg/util/subnet/const.go [hunk 7]
```diff
@@ -5,6 +5,5 @@ package subnet
 
 // Network security group constants
 const (
-	NSGControlPlaneSuffix = "-controlplane-nsg"
-	NSGNodeSuffix         = "-node-nsg"
+	NSGSuffix = "-nsg"
```
Comment (jim-minter) [inline pkg/util/subnet/const.go:8]: The changes in pkg/util/subnet are going to break dynamic validation for clusters on the <4.5 architecture.  This is going to need more thought.

File: hack/vendor/vendor.go (hunk 1 above)
Comment (jim-minter) [reply hack/vendor/vendor.go:1]: Would our lives not be radically simpler if we had a carry patch on the installer that just checked in assets_vfsdata.go?

File: pkg/util/clusterdata/clusterdata.go [hunk 8]
```diff
@@ -3,7 +3,6 @@ package clusterdata
 // Copyright (c) Microsoft Corporation.
 // Licensed under the Apache License 2.0.
 
-// TODO: After OpenShift 4.4, replace github.com/openshift/cluster-api with github.com/openshift/machine-api-operator
```
[the remark this answers is not in the document]
Comment (jim-minter) [reply pkg/util/clusterdata/clusterdata.go:6]: Ah, there are two cases of this - another in the monitor.

File: Makefile (hunk 2 above)
Comment (jim-minter) [reply Makefile:114]: For now: we'll try to get rid of vendor.go, keep vendor/, add -mod=vendor

File: pkg/util/subnet/const.go (hunk 7 above)
Comment (jim-minter) [reply pkg/util/subnet/const.go:8]: i.e. the problem is NetworkSecurityGroupID

Comment (mjudeikis) [reply pkg/util/subnet/const.go:8]: We need backward compatibility for NSG and LB for all the functions to cover both architectures.

Comment (mjudeikis) [reply pkg/util/subnet/const.go:8]: Options: Architecture version flag in the documents, where we version architectures. And some kind of function where we decide what is what

File: go.mod [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+	github.com/metal3-io/baremetal-operator => github.com/openshift/baremetal-operator v0.0.0-20200611190251-d997d9c06ba8
+	github.com/metal3-io/cluster-api-provider-baremetal => github.com/openshift/cluster-api-provider-baremetal v0.0.0-20190821174549-a2a477909c1d
+	github.com/openshift/api => github.com/openshift/api v0.0.0-20200413201024-c6e8c9b6eb9a
+	github.com/openshift/client-go => github.com/openshift/client-go v0.0.0-20200521150516-05eb9880269c
+	github.com/openshift/installer => github.com/jim-minter/installer v0.0.0-20200915125021-47a8a781f957
+	github.com/openshift/machine-api-operator => github.com/openshift/machine-api-operator v0.0.0-20200529045911-d19e8d007f7c
+	github.com/openshift/machine-config-operator => github.com/openshift/machine-config-operator v0.0.0-20200904000724-41d29dde06d6 // ARO Carry-on: to avoid k8s version miss-align
+	github.com/vmware/govmomi => github.com/vmware/govmomi v0.22.2-0.20200420222347-5fceac570f29
+	google.golang.org/api => google.golang.org/api v0.13.0
+	k8s.io/api => k8s.io/api v0.18.3
```
[the remark this answers is not in the document]
Comment (mjudeikis) [reply go.mod:86]: DIscussed this with some of the upstream component owners and was persuaded not to do this due to not aligned vendoring strategy and "random fixes" in our upstream branches, which might not be suitable for us. Happy to discuss in private.

File: hack/vendor/vendor.go (hunk 1 above)
Comment (mjudeikis) [reply hack/vendor/vendor.go:1]: removed

File: docs/publish-rhcos-image.md (hunk 4 above)
Comment (mjudeikis) [reply docs/publish-rhcos-image.md:7]: not anymore :)

**Comment by jim-minter** [comment]
@mjudeikis I've rebased this, done some work to sanitise the commits, removed the V->v commit and added a trial workaround patch for the LB/PLS bug. Your old branch is at https://github.com/jim-minter/ARO-RP/tree/mjudeikis/release-4.5-old in case you want to diff.

Some outstanding tasks that need to be done before merge:

* refactor installer/release-4.5-azure to sanitise commits [MJ]
* review usage of context.TODO(), context.Background() [either]
* review go.mod settings for sanity - validate we're bringing in the right releases of things [JM]

@hawkowl please also review this PR.

File: hack/update-go-module-dependencies.sh [hunk 10]
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash -ex
```
Comment (mjudeikis) [inline hack/update-go-module-dependencies.sh:1]: This script looks terrible :/

Comment (jim-minter) [reply hack/update-go-module-dependencies.sh:1]: I like to think that it's not pretending to be something that it's not.

**Comment by jim-minter** [comment]
successful e2e. running again.

**Comment by jim-minter** [comment]
successful e2e. running again.

**Review by jim-minter — APPROVED** [review]
Let's do it.

**Pull request merged**